### PR TITLE
Fixed linting errors and warnings

### PR DESCRIPTION
**Pull request by janechu**
# Pull Request

## 📖 Description

This change does the following:
- Allow `no-non-null-assertion`
- Fixes all other warnings and errors

## 👩‍💻 Reviewer Notes

@EisenbergEffect since these changes affect your work please take a close look, most of the changes are benign but I did move code around to best minimize ignoring the no-use-before-define issues.

### General

- [ ] I have included a change request file using `$ yarn change`
- [ ] I have added tests for my changes.
- [x] I have tested my changes.
- [ ] I have updated the project documentation to reflect my changes.
- [x] I have read the [CONTRIBUTING](https://github.com/Microsoft/fast/blob/master/CONTRIBUTING.md) documentation and followed the [standards](https://www.fast.design/docs/community/code-of-conduct/#our-standards) for this project.

### CODE REVIEW

File: packages/web-components/fast-element/src/templating/binding.ts [hunk 1]
```diff
@@ -33,7 +33,7 @@ export interface BindingMode {
     event: BindingType;
 }
 
-export interface BindingConfig<T = any> {
+export interface BindingConfig {
```
Comment (EisenbergEffect) [inline packages/web-components/fast-element/src/templating/binding.ts:36]: I think we need this generic for inferred types in actual usage scenarios. Possible to add it back?

**Review by EisenbergEffect — CHANGES_REQUESTED** [review]

File: packages/web-components/fast-element/src/templating/binding.ts (hunk 1 above)
Comment (janechu) [reply packages/web-components/fast-element/src/templating/binding.ts:36]: I've got a warning/error that it was unused, I can add it back and change the eslint config instead.

Comment (janechu) [reply packages/web-components/fast-element/src/templating/binding.ts]: Added it as a one line ignore.

File: packages/web-components/fast-element/src/templating/binding.ts [hunk 2]
```diff
@@ -62,12 +63,158 @@ class BindingBase {
...
 class TargetUpdateBinding extends BindingBase {
     constructor(directive: HTMLBindingDirective, protected updateTarget: UpdateTarget) {
         super(directive);
     }
 
-    static createBindingConfig<T>(defaultOptions: T, eventType?: BindingType) {
+    static createBindingConfig<T>(
+        defaultOptions: T,
+        eventType?: BindingType
+    ): BindingConfig {
```
Comment (EisenbergEffect) [inline packages/web-components/fast-element/src/templating/binding.ts:217]: Needs to be `BindingConfig<T>`

**Review by EisenbergEffect — APPROVED** [review]
Just need that one more generic to pass through I think.

**Pull request merged**